### DISCUSSION TITLE
为什么TCP挥手时主动发送方需要linger？

### DISCUSSION THREAD

**Discussion by WellOptimized** [liusy58/WhyThis#37] [General]
在本文中称主动发起挥手的一方为client，另一方为server。

在说到TCP的四次挥手时，经常会提到[两军问题](https://en.wikipedia.org/wiki/Two_Generals%27_Problem)，TCP为了十分接近的解决两军问题，采用了client进入`TIME-WAIT`的方法。

归纳一下client或者server进入`clean closed`的**先决条件**：

1. 收到`FIN`并且收到的所有段已经组装完成。
2. 将`FIN`发送出去并且local application不再产生数据。
3. 发送的segment都被确认。
4. 确信remote peer满足条件3。

![image](https://user-images.githubusercontent.com/38347398/119262412-bd369600-bc0d-11eb-9aed-e096eb48c5e7.png)

列出四次挥手的流程：

1. client将`FIN`发送出去并且local application不再产生数据。

   client达成条件（2）
   server达成条件（1）

2. server对`FIN`进行确认。

   client达成条件（2，3）
   server达成条件（1，4）

   server是如何确信client满足（3）的？client没有重传`FIN`。

3. server将`FIN`发送出去并且local application不再产生数据。

   client达成条件（1，2，3）
   server达成条件（1，2，4）

4. client对`FIN`进行确认。

   在这里client处于`TIME-WAIT`，然后观察client和server缺失的条件，发现client想要达成的条件（4）依赖于server想要达成的条件（3）。那client和server的达成条件过程是

   1. client（1，2，3）；server（1，2，4）  

   2. client（1，2，3）；server（1，2，3，4）

      server想要完成从4.1到4.2的转换，需要接收client发送的`ACK`。

   3.  client（1，2，3，4）；server（1，2，3，4）

      client想要完成从4.2到4.3的转换，需要server不重传包，也就是server接收`ACK`。

   再回看一下4.2和4.3，发现它们的先决条件是一样的。所以只要server接收到client的`ACK`，整个挥手过程就完成了。然而因为网络中的各种因素，server有时没办法一次就接收到client的`ACK`，需要设定一段时间来给server重传`FIN`，这段时间也就是client处于`linger`的时间2MSL。2MSL= 去向`ACK`消息最大存活时间（MSL) + 来向`FIN`消息的最大存活时间(MSL)。 如果过了2MSL，client没有收到重传的`FIN`，就可以默认server接收了`ACK`，整个过程就顺利结束。如果在2MSL内client收到重传的FIN，那就发送`ACK`并重新2MSL的计时。

**Comment by liusy58** [comment]
我认为是不是加一些图更好理解？
